### PR TITLE
print why a test was ignored if its the only test specified

### LINKED ISSUE #106659: compiletest: Print why a test was ignored if it's the only test specified
cc https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/.E2.9C.94.20bless.20src.2Ftest.2Fui.2Fstats.2Fhir-stats.2Ers.3F, where it took a while to track down why the test was ignored. Doing this for all tests is too much, because there can be many ignored tests in a full `x.py test` run, but when only a single test is specified with `x test src/test/ui/stats/hir-stats.rs` it's pretty clear that the user intends to run it.

I think https://github.com/rust-lang/rust/blob/b435960c4cfd3975651c7051be56d7f5d6c201ab/src/tools/compiletest/src/header.rs#L951-L975 is roughly the right part of the code to change.

### PR DESCRIPTION
**Pull request by lenko-d**
Fixes [#106659](https://github.com/rust-lang/rust/issues/106659)

### CODE REVIEW

File: library/test/src/lib.rs [hunk 1]
```diff
@@ -514,6 +516,21 @@ pub fn filter_tests(opts: &TestOpts, tests: Vec<TestDescAndFn>) -> Vec<TestDescA
     filtered
 }
 
+/// Custom handling of cases where there is only 1 test to be executed.
+/// This is needed because in such cases we may want to show more detailed information.
+/// For example for investigation purposes we may want to show the reasons why the test case was ignored.
+fn check_for_single_test_run(filtered_tests: &Vec<TestDescAndFn>) {
```
Comment (Nilstrieb) [inline library/test/src/lib.rs:522]: These changes to libtest also seem useful, but could you move them to a separate PR and only change compiletest here?

File: src/librustdoc/doctest.rs [hunk 2]
```diff
@@ -1069,6 +1069,7 @@ impl Tester for Collector {
                     Ignore::Some(ref ignores) => ignores.iter().any(|s| target_str.contains(s)),
                 },
                 ignore_message: None,
+                ignore_reasons: None,
```
Comment (Nilstrieb) [inline src/librustdoc/doctest.rs:1072]: I think you forgot to move this out as well

File: src/tools/compiletest/src/header.rs [hunk 3]
```diff
@@ -934,48 +934,175 @@ pub fn make_test_description<R: Read>(
...
+
+        ignore |= reason("ignore_llvm", ignore_llvm(config, ln), ignore, &mut ir);
+        ignore |= reason(
+            "run_clang_based_tests_with.is_none && parse_needs_matching_clang",
+            config.run_clang_based_tests_with.is_none() && config.parse_needs_matching_clang(ln),
+            ignore,
+            &mut ir,
+        );
+        ignore |= reason(
+            "!has_asm_support && needs-asm-support",
```
Comment (Mark-Simulacrum) [inline src/tools/compiletest/src/header.rs:982]: I think it would be much cleaner in this case if we made reason a macro, and used stringify! to generate the ignore reason. That'll be good enough (especially for internal usage like compiletest) and avoid bugs in this. It also means we can skip passing ignore and &mut ir; they can be implicit, I think.

**Comment by Mark-Simulacrum** [comment]
r=me with commits squashed, thanks!
